[HOLD for PR #42892] trim leading and trailing CR LF during paste into composer

kpadmanabhan opened the pull request:
@jayeshmangwani / @roryabraham 

### Details
Trimming leading and trailing CR LF characters from pasted content of composer text, so that it does not hamper the flow of `paste` event triggering and handling in Newdot composer.

### Fixed Issues
$ https://github.com/Expensify/App/issues/39971
PROPOSAL: https://github.com/Expensify/App/issues/39971#issuecomment-2053860349


### Tests
1. Make sure composer field is empty
2. Copy below mentioned text(s) into clipboard
3. Paste it in composer and check if "Send" button is enabled and clickable

Text(s) tested: 

- [ ] Single line without trailing CR LF

```
This is a single line text.
```

- [ ] Single line with trailing CR LF

```
This is single line text but with newline at the end of line.

```

- [ ] Single line with leading CR LF

```

This is single line text but with newline at the start of line.
```

- [ ] Single line with leading and trailing CR LF

```

This is single line text but with newline at the start and end of line.

```

- [ ] Text with newline character in text content

```
Text with newline character \n in text content.\n
```

- [ ] Multi line without trailing CR LF

```
During communication with Expensify, you will come across a variety of acronyms used by our team. While acronyms can be
 useful, they cease to be the moment they are not known to the receiver. As such, we wanted to create a list here of our most 
commonly used acronyms and what they're referring to. Lastly, please never hesitate to ask in Slack or the GH issue if there are 
any that are not understood/known!
```

- [ ] Multi line with trailing CR LF

```
During communication with Expensify, you will come across a variety of acronyms used by our team. While acronyms can be 
useful, they cease to be the moment they are not known to the receiver. As such, we wanted to create a list here of our most 
commonly used acronyms and what they're referring to. Lastly, please never hesitate to ask in Slack or the GH issue if there are 
any that are not understood/known!

```

- [ ] Numbered text

```
    1. Proposing steps to take for a regression test to ensure the bug doesn't occur again (For information on how to successfully complete this, head [here](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md)).
    2. Identifying and noting the offending PR that caused the bug (if any).
    3. Commenting on the offending PR to note the bug it caused and why (if applicable).
    4. Starting a conversation on if any additional steps should be taken to prevent further bugs similar to the one fixed from occurring again. 
    5. 
```

- [ ] Code comment

```
    ```
    [commit]
        gpgsign = true
    [user]
        email = <Your GH account email>
        name = <Your Name>
        signingkey = <your_signing_key>
    [gpg]
        program = gpg
    ```
```

- [ ] Bulleted text

```
- If you have made a change to your pull request and are ready for another review, leave a comment that says "Updated" on the pull request  itself.
- Please keep the conversation in GitHub, and do not ping individual reviewers in Slack or Upwork to get their attention.
- Pull Request reviews can sometimes take a few days. If your pull request has not been addressed after four days, please let us know via the #expensify-open-source Slack channel.
```

- [ ] Html text

```
<input onpaste={(event:ClipboardEvent<HTMLInputElement>) => pasteMock(event)} />
```

- [ ] Multi-line html text
```
<div align="center">
    <a href="https://new.expensify.com">
        <img src="https://raw.githubusercontent.com/Expensify/App/main/web/favicon.png" width="64" height="64" alt="New Expensify Icon">
    </a>
    <h1>
        <a href="https://new.expensify.com">
            New Expensify
        </a>
    </h1>
</div>

```

- [ ] Copy paste an image

- [ ] Verify that no errors appear in the JS console

### Offline tests
Same as above.

### QA Steps
Same as above.
Any pasted text (copy to clipboard and paste it into the composer) containing leading or trailing newline Carriage Return (CR) and Line Feed (LF) characters must enable "Send" button of composer.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


https://github.com/Expensify/App/assets/76063588/3cee172b-64b7-413e-9aa5-d12592a9fcb9



</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/76063588/56af88f7-b7a6-4c24-abbd-d212c8e9edf0



</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/76063588/6efe0332-cc87-44da-a487-3c3664033d50


</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/Expensify/App/assets/76063588/bf331096-81b1-465e-8157-1f62644b31ae



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/76063588/b4214604-48f5-42e3-8387-396de246c2f5



</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/76063588/c6b0834f-047a-4df2-b277-998fffb77e7e


</details>

kpadmanabhan committed:
trim leading and trailing CR LF during paste into composer

jayeshmangwani commented:
Hey @kpadmanabhan , Are you able to run the app on all the platforms? Please fill in all the missing checkboxes from the description that can be finished. And please write the unit test for the useHtmlPaste function.

kpadmanabhan commented:
@jayeshmangwani  Even though my change is not OS dependent, I understand that I must test it in all env. I was using windows till now. Just got hold of a mac. Setting up env in that. Whom can I contact for a Mapbox key?

jayeshmangwani commented:
@kpadmanabhan For the key, I have responded in Slack. If you need any help in setting it up, please hit me up.

jayeshmangwani commented:
Hey @kpadmanabhan, is there any update for us? If you are stuck anywhere, you can comment here or [on #expensify-open-source](https://expensify.enterprise.slack.com/archives/C01GTK53T8Q).

jayeshmangwani commented:
@kpadmanabhan bump for the update

roryabraham commented:
Seems like this PR is abandoned

jayeshmangwani commented:
Yes @roryabraham , it looks like the contributor is not active

kpadmanabhan commented:
@roryabraham  / @jayeshmangwani : apologies for this. I was quite unwell for past couple of weeks. Sorry for no updates. Resuming work today. Fixes should be addressed by today e.o.d.

kpadmanabhan commented:
Testing on iOS (native and web), macOS (chrome web and desktop app) complete. Results attached to PR.
Facing challenges with Android testing on Pixel simulator. Though script brings up android simulator, it executes and then stops the emulator by itself before launching the app with no visible error messages in console.

Where can I find detailed logging on whats going wrong with the launch of simulator and starting the app?

roryabraham commented:
I find I often have to delete the emulator and create new ones. Or if you have access to a physical Android device I find that to be much more reliable than emulators.

kpadmanabhan commented:
Now getting below error. 
@roryabraham / @jayeshmangwani  Any thioughts?

```
Warning: SDK processing. This version only understands SDK XML versions up to 3 but an SDK XML file of version 4 was encountered. This can happen if you use versions of Android Studio and the command-line tools that were released at different times.
Warning: SDK processing. This version only understands SDK XML versions up to 3 but an SDK XML file of version 4 was encountered. This can happen if you use versions of Android Studio and the command-line tools that were released at different times.
CMake Deprecation Warning at CMakeLists.txt:42 (cmake_policy):
  The OLD behavior for policy CMP0026 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


CMake Warning at /Users/archanavenkat/Library/Android/sdk/cmake/3.22.1/share/cmake-3.22/Modules/Platform/Darwin-Initialize.cmake:303 (message):
  Ignoring CMAKE_OSX_SYSROOT value:

   /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk

  because the directory does not exist.
Call Stack (most recent call first):
  /Users/archanavenkat/Library/Android/sdk/cmake/3.22.1/share/cmake-3.22/Modules/CMakeSystemSpecificInitialize.cmake:21 (include)
  CMakeLists.txt:64 (project)


CMake Error at /Users/archanavenkat/Library/Android/sdk/cmake/3.22.1/share/cmake-3.22/Modules/CMakeTestCCompiler.cmake:69 (message):
  The C compiler

    "/Library/Developer/CommandLineTools/usr/bin/cc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /Users/archanavenkat/krishnan/App/node_modules/react-native/ReactAndroid/hermes-engine/build/hermes/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/make -f Makefile cmTC_91bc9/fast && /Applications/Xcode.app/Contents/Developer/usr/bin/make  -f CMakeFiles/cmTC_91bc9.dir/build.make CMakeFiles/cmTC_91bc9.dir/build
    Building C object CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o
    /Library/Developer/CommandLineTools/usr/bin/cc   -arch arm64 -mmacosx-version-min=11.3 -MD -MT CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o -MF CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o.d -o CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o -c /Users/archanavenkat/krishnan/App/node_modules/react-native/ReactAndroid/hermes-engine/build/hermes/CMakeFiles/CMakeTmp/testCCompiler.c
    Linking C executable cmTC_91bc9
    /Users/archanavenkat/Library/Android/sdk/cmake/3.22.1/bin/cmake -E cmake_link_script CMakeFiles/cmTC_91bc9.dir/link.txt --verbose=1
    /Library/Developer/CommandLineTools/usr/bin/cc  -arch arm64 -mmacosx-version-min=11.3 -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o -o cmTC_91bc9 
    ld: library 'System' not found
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    make[1]: *** [cmTC_91bc9] Error 1
    make: *** [cmTC_91bc9/fast] Error 2
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:64 (project)



FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':react-native:packages:react-native:ReactAndroid:hermes-engine:configureBuildForHermes'.
> Process 'command '/Users/archanavenkat/Library/Android/sdk//cmake/3.22.1/bin/cmake'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 7m 4s
error Failed to install the app. Command failed with exit code 1: ./gradlew app:installDevelopmentDebug -PreactNativeDevServerPort=8081
Warning: SDK processing. This version only understands SDK XML versions up to 3 but an SDK XML file of version 4 was encountered. This can happen if you use versions of Android Studio and the command-line tools that were released at different times.
Warning: SDK processing. This version only understands SDK XML versions up to 3 but an SDK XML file of version 4 was encountered. This can happen if you use versions of Android Studio and the command-line tools that were released at different times.
CMake Deprecation Warning at CMakeLists.txt:42 (cmake_policy): The OLD behavior for policy CMP0026 will be removed from a future version of CMake. The cmake-policies(7) manual explains that the OLD behaviors of all policies are deprecated and that a policy should be set to OLD only under specific short-term circumstances. Projects should be ported to the NEW behavior and not rely on setting a policy to OLD. CMake Warning at /Users/archanavenkat/Library/Android/sdk/cmake/3.22.1/share/cmake-3.22/Modules/Platform/Darwin-Initialize.cmake:303 (message): Ignoring CMAKE_OSX_SYSROOT value: /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk because the directory does not exist.
Call Stack (most recent call first): /Users/archanavenkat/Library/Android/sdk/cmake/3.22.1/share/cmake-3.22/Modules/CMakeSystemSpecificInitialize.cmake:21 (include) CMakeLists.txt:64 (project) CMake Error at /Users/archanavenkat/Library/Android/sdk/cmake/3.22.1/share/cmake-3.22/Modules/CMakeTestCCompiler.cmake:69 (message): The C compiler "/Library/Developer/CommandLineTools/usr/bin/cc" is not able to compile a simple test program. It fails with the following output: Change Dir: /Users/archanavenkat/krishnan/App/node_modules/react-native/ReactAndroid/hermes-engine/build/hermes/CMakeFiles/CMakeTmp Run Build Command(s):/usr/bin/make -f Makefile cmTC_91bc9/fast && /Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/cmTC_91bc9.dir/build.make CMakeFiles/cmTC_91bc9.dir/build Building C object CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o /Library/Developer/CommandLineTools/usr/bin/cc -arch arm64 -mmacosx-version-min=11.3 -MD -MT CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o -MF CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o.d -o CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o -c /Users/archanavenkat/krishnan/App/node_modules/react-native/ReactAndroid/hermes-engine/build/hermes/CMakeFiles/CMakeTmp/testCCompiler.c Linking C executable cmTC_91bc9 /Users/archanavenkat/Library/Android/sdk/cmake/3.22.1/bin/cmake -E cmake_link_script CMakeFiles/cmTC_91bc9.dir/link.txt --verbose=1 /Library/Developer/CommandLineTools/usr/bin/cc -arch arm64 -mmacosx-version-min=11.3 -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/cmTC_91bc9.dir/testCCompiler.c.o -o cmTC_91bc9 ld: library 'System' not found clang: error: linker command failed with exit code 1 (use -v to see invocation) make[1]: *** [cmTC_91bc9] Error 1 make: *** [cmTC_91bc9/fast] Error 2 CMake will not be able to correctly generate this project.
Call Stack (most recent call first): CMakeLists.txt:64 (project) FAILURE: Build failed with an exception. * What went wrong:
Execution failed for task ':react-native:packages:react-native:ReactAndroid:hermes-engine:configureBuildForHermes'.
> Process 'command '/Users/archanavenkat/Library/Android/sdk//cmake/3.22.1/bin/cmake'' finished with non-zero exit value 1 * Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org. BUILD FAILED in 7m 4s.
info Run CLI with --verbose flag for more details.

```

jayeshmangwani commented:
@kpadmanabhan I haven't gotten the above issue, but I got an issue on Android that was solved by `removing the android/app/.cxx folder and then re-run`

jayeshmangwani commented:
@kpadmanabhan Any luck with running on Android ?

If you still facing the issue, then I would suggest leave the Android And Android mWeb testing for me, I will take care of it, You can focus on remaining part of the PR,  And then you can fix your Android environment later.

kpadmanabhan commented:
@jayeshmangwani : thanks for your quick support to investigate the setup. I have fixed the setup and am able to run in Android as well. Updated results of all in the template. Kindly review.

jayeshmangwani commented:
@kpadmanabhan You need to sign the all commits, and lint code is failing here FIX that, remove the commit from a user name `Archana Venkataraman`

jayeshmangwani commented on src/hooks/useHtmlPaste/index.ts:
```diff
@@ -12,6 +12,7 @@ const insertAtCaret = (target: HTMLElement, text: string) => {
     if (selection?.rangeCount) {
         const range = selection.getRangeAt(0);
         range.deleteContents();
+        text = text.replace(/^[\r\n]+|[\r\n]+$/g, '');  // Remove trailing Carriage Return and Line feed
```
Define this regex in the [CONST](https://github.com/Expensify/App/blob/8375abea3a377e7416404addde255b6ac8530a9b/src/CONST.ts#L1894) file

kpadmanabhan replied:
Fixed

kpadmanabhan committed:
Merge branch 'main' into 39971_sanitize_pasted_text_CRLF

kpadmanabhan committed:
fixing review comment

kpadmanabhan committed:
lint fix

kpadmanabhan commented:
@jayeshmangwani : can you pls re-review this PR?

jayeshmangwani commented on src/hooks/useHtmlPaste/index.ts:
```diff
@@ -12,7 +13,8 @@ const insertAtCaret = (target: HTMLElement, text: string) => {
         const range = selection.getRangeAt(0);
         range.deleteContents();
-        const node = document.createTextNode(text);
+        const sanitizedText = text.replace(CONST.REGEX.TRAILING_CRLF, '');  // Remove trailing Carriage Return and Line feed
```
```suggestion
        const sanitizedText = text.replace(CONST.REGEX.TRAILING_CRLF, ''); // Remove trailing Carriage Return and Line feed
```

jayeshmangwani commented:
@kpadmanabhan Can we please add the unit test for useHtmlPaste ?

jayeshmangwani commented:
Locally, lint is failing for me for some reason. Does lint pass for you?

kpadmanabhan commented:
> Locally, lint is failing for me for some reason. Does lint pass for you?

yes. lint is passing for me locally. did you pull latest?
```
(base) XXX@XX-MacBook-Air App % npm run lint

> new.expensify@1.4.75-0 lint
> eslint . --max-warnings=0 --cache --cache-location=node_modules/.cache/eslint

(base) XXX@XX-MacBook-Air App % 
```

kpadmanabhan committed:
Update src/hooks/useHtmlPaste/index.ts

spacing fix

Co-authored-by: Jayesh Mangwani <35371050+jayeshmangwani@users.noreply.github.com>

jayeshmangwani commented:
> yes. lint is passing for me locally. did you pull latest?

Yes, it tested in the latest pull; maybe it's only for me

kpadmanabhan committed:
added unit tests

kpadmanabhan committed:
merge from remote

kpadmanabhan commented:
@jayeshmangwani : added unit tests in StringUtils as this is a regex check. Pls re-review.

jayeshmangwani commented:
Thanks @kpadmanabhan for adding the unit tests; I was testing this PR and one scenario I missed before, and I want to confirm before we merge the PR.

Right now, in the main branch in composer, when we copy the multi newline and paste it between the large message, then new lines don't lose the originally copied newlines; and now, after adding the above changes in this PR, when we copy the muli new lines text between the large message then leading and trailing newlines got removed, below I am adding two videos for the visual effect of the changes between main and this PR
cc: @roryabraham 

<details>
  <summary>Main branch</summary>

https://github.com/Expensify/App/assets/35371050/53495222-934d-4509-b987-77bd3ece91b8


</details>

<details>
  <summary>Existing PR</summary>
  
  

https://github.com/Expensify/App/assets/35371050/6b3f5b55-9bbb-4ae7-81ea-5da2fbe85482


</details>

kpadmanabhan commented:
> Thanks @kpadmanabhan for adding the unit tests; I was testing this PR and one scenario I missed before, and I want to confirm before we merge the PR.
> 
> Right now, in the main branch in composer, when we copy the multi newline and paste it between the large message, then new lines don't lose the originally copied newlines; and now, after adding the above changes in this PR, when we copy the muli new lines text between the large message then leading and trailing newlines got removed, below I am adding two videos for the visual effect of the changes between main and this PR cc: @roryabraham
> 
> Main branch
>  new-lines-main.mov 
> Existing PR
>  new-lines-after-this-PR.mov

@jayeshmangwani : thanks for this test. 
the reason for this behaiour is that in the hooks when we paste a text, we only have access to the pasted text and not the text already present in the composer. if the trailing newlines are not removed, then the event does not get propagated to composer, resulting in send button not getting enabled.

> ONYX update in `ComposerWithSuggestions.tsx` is not fired through `onChangeText` event and hence draft report is not saved to ONYX.
> This is because `isCommentEmpty` returns TRUE.

jayeshmangwani commented:
>  if the trailing newlines are not removed, then the event does not get propagated to composer, resulting in send button not getting enabled.

There must be something wrong with the existing code event that should be fired

kpadmanabhan commented:
> > if the trailing newlines are not removed, then the event does not get propagated to composer, resulting in send button not getting enabled.
> 
> There must be something wrong with the existing code event that should be fired

the reason being that start and end of selection range from the node are detected as 0 due to the presence of trailing newline and hence propagation of event does not happen to composer `handlePaste()`.

roryabraham commented:
The last bug reported by @jayeshmangwani does look like a regression that we have to address before we can merge this PR.

> the reason for this behaiour is that in the hooks when we paste a text, we only have access to the pasted text and not the text already present in the composer

Can we pass the text already present in the composer as a param to the hook?

jayeshmangwani commented:
@kpadmanabhan @roryabraham Let's close/HOLD this PR. [This proposal](https://github.com/Expensify/App/issues/42216#issuecomment-2115186516) was selected for another issue, and that proposal will solve [our original](https://github.com/Expensify/App/issues/39971) issue.

kpadmanabhan commented:
> Right now, in the main branch in composer, when we copy the multi newline and paste it between the large message, then new lines don't lose the originally copied newlines; and now, after adding the above changes in this PR, when we copy the muli new lines text between the large message then leading and trailing newlines got removed, below I am adding two videos for the visual effect of the changes between main and this PR cc: @roryabraham

@jayeshmangwani / @roryabraham : i have fixed the issue you reported by checking for existing hooks
however, if there is a newline at the end in the composer, I see errors in console and am fixing that. I think as https://github.com/Expensify/App/issues/42216#issuecomment-2115186516 has not created a PR until now, it is better to go with this current PR.

jayeshmangwani commented:
> I think as https://github.com/Expensify/App/issues/42216#issuecomment-2115186516 has not created a PR until now, it is better to go with this current PR.

Technically, we should go with existing PR as this is already in motion, but our changes of replacing the new lines (CR LF) are causing a regression, and the solution on the other issue looks good and does not cause any known regression, so I do think we should go with the other proposal

kpadmanabhan commented:
@jayeshmangwani / @roryabraham : kindly suggest whether I should continue with this PR or put it on hold until other PR is deployed. In which case what will happen to the effort and time spent on this PR?

kpadmanabhan commented:
> @jayeshmangwani / @roryabraham : kindly suggest whether I should continue with this PR or put it on hold until other PR is deployed. In which case what will happen to the effort and time spent on this PR?

@jayeshmangwani / @roryabraham : how should I continue with this? kindly recommend.

jayeshmangwani commented:
> kindly suggest whether I should continue with this PR or put it on hold until other PR is deployed

@kpadmanabhan HOLD this PR please, our original proposed changes will cause the regression and other [PR](https://github.com/Expensify/App/pull/42892) is close to getting merged 

> In which case what will happen to the effort and time spent on this PR?

I will raise this internally and will post here once I get the update

cc: @roryabraham

jayeshmangwani commented:
@kpadmanabhan It looks like the issue has been fixed now. I am no longer able to reproduce it. Can you please verify?

kpadmanabhan commented:
> @kpadmanabhan It looks like the issue has been fixed now. I am no longer able to reproduce it. Can you please verify?

@jayeshmangwani : 
https://github.com/Expensify/App/pull/42892#issuecomment-2227120479

jayeshmangwani commented:
@kpadmanabhan Please close this PR

kpadmanabhan commented:
> @kpadmanabhan Please close this PR

This issue is not reproducible in main HEAD anymore. Closing this PR.
